csrc/id_model/indexing.cpp

```diff
@@ -1042,12 +1042,19 @@ std::unordered_map<Val*, Val*> TensorIndexer::getIndexReplacementMap(
         // If this for-loop is a circular buffer loop, the loop index
         // may need to have an additional offset
         if (!as_consumer) {
+          Val* base_index =
+              replacement_index != nullptr ? replacement_index : cur_index;
           if (auto circular_buffer_offset =
                   getLoopIndexOffsetForProducerOfCircularBuffer(
                       expr, for_loop, id_model_)) {
             replacement_index = SimplifyingIrBuilder::addExpr(
-                replacement_index != nullptr ? replacement_index : cur_index,
+                for_loop->isTrivial() ? for_loop->start() : base_index,
                 circular_buffer_offset);
+          } else if (
+              for_loop->circularBufferLoopStage() !=
+              CircularBufferLoopStage::NotApplicable) {
+            replacement_index =
```

Comment:
Isn't this replacement necessary always, no matter if it's for consumer or producer indexing?

Reply:
If so, I wonder why nothing breaks without the replacement? Since the loop is trivial, no loop is generated, but without the replacement, the index keeps using the loop index, which should result in a compilation error, no?

Reply:
I updated the PR to always do the replacement.

Reply:
Since we're only doing circular buffering with TMA loads, it is always the producer index into global memory.
The consumer index for shared memory always starts from the base index.

Reply:
Since id-model indexing supports TMA circular buffers, I'll refactor some of my manual indexing.